File: Samples/Evaluate/winforms-spreadsheet-create-custom-progress-indicator-21.1.3-/CS/SpreadsheetProgressSample/WaitForm1.cs
using System;
using System.Threading;
using DevExpress.XtraWaitForm;

namespace SpreadsheetProgressSample {
    public partial class WaitForm1 : WaitForm {
        CancellationTokenSource cancellationTokenSource;

        public WaitForm1() {
            InitializeComponent();
            progressPanel1.AutoSize = true;
        }

        public override void SetCaption(string caption) {
            base.SetCaption(caption);
            progressPanel1.Caption = caption;
        }

        public override void SetDescription(string description) {
            base.SetDescription(description);
            progressPanel1.Description = description;
        }

        public override void ProcessCommand(Enum cmd, object arg) {
            base.ProcessCommand(cmd, arg);
            WaitFormCommand command = (WaitFormCommand)cmd;
            if (command == WaitFormCommand.SetCancellationTokenSource)
                cancellationTokenSource = (CancellationTokenSource)arg;
        }

        public enum WaitFormCommand {
            SetCancellationTokenSource
        }

        void Cancel_Click(object sender, EventArgs e) {
            cancellationTokenSource?.Cancel();
        }
    }
}

File: Samples/Evaluate/winforms-spreadsheet-create-custom-progress-indicator-21.1.3-/VB/SpreadsheetProgressSample/WaitForm1.vb
Imports System
Imports System.Threading
Imports DevExpress.XtraWaitForm

Namespace SpreadsheetProgressSample
	Partial Public Class WaitForm1
		Inherits WaitForm

		Private cancellationTokenSource As CancellationTokenSource

		Public Sub New()
			InitializeComponent()
			progressPanel1.AutoSize = True
		End Sub

		Public Overrides Sub SetCaption(ByVal caption As String)
			MyBase.SetCaption(caption)
			progressPanel1.Caption = caption
		End Sub

		Public Overrides Sub SetDescription(ByVal description As String)
			MyBase.SetDescription(description)
			progressPanel1.Description = description
		End Sub

		Public Overrides Sub ProcessCommand(ByVal cmd As System.Enum, ByVal arg As Object)
			MyBase.ProcessCommand(cmd, arg)
			Dim command As WaitFormCommand = CType(cmd, WaitFormCommand)
			If command = WaitFormCommand.SetCancellationTokenSource Then
				cancellationTokenSource = DirectCast(arg, CancellationTokenSource)
			End If
		End Sub

		Public Enum WaitFormCommand
			SetCancellationTokenSource
		End Enum

		Private Sub Cancel_Click(ByVal sender As Object, ByVal e As EventArgs) Handles hyperlinkLabelControl1.Click
			cancellationTokenSource?.Cancel()
		End Sub
	End Class
End Namespace